src/util/walkController.ts

```diff
@@ -0,0 +1,132 @@
+import { BaseTrie } from '..'
+import { FoundNodeFunction } from '../baseTrie'
+import { PrioritizedTaskExecutor } from '../prioritizedTaskExecutor'
+import { BranchNode, ExtensionNode, LeafNode, Nibbles, TrieNode } from '../trieNode'
+
+/**
+ * WalkController is an interface to control how the trie is being traversed.
+ */
+export default class WalkController {
+  readonly onNode: FoundNodeFunction
+  readonly taskExecutor: PrioritizedTaskExecutor
+  readonly trie: BaseTrie
+  private resolve: Function
+
+  /**
+   * Creates a new WalkController
+   * @param onNode - The `FoundNodeFunction` to call if a node is found
+   * @param trie - The `Trie` to walk on
+   */
+  private constructor(onNode: FoundNodeFunction, trie: BaseTrie, poolSize: number) {
+    this.onNode = onNode
+    this.taskExecutor = new PrioritizedTaskExecutor(poolSize)
+    this.trie = trie
+    this.resolve = () => {}
+  }
+
+  /**
+   * Async function to create and start a new walk over a trie.
+   * @param onNode - The `FoundNodeFunction to call if a node is found
+   * @param trie - The trie to walk on
+   * @param root - The root key to walk on
+   * @param poolSize - Task execution pool size to prevent OOM errors. Defaults to 500.
+   */
+  static async newWalk(
+    onNode: FoundNodeFunction,
+    trie: BaseTrie,
+    root: Buffer,
+    poolSize?: number
+  ): Promise<void> {
+    const strategy = new WalkController(onNode, trie, poolSize ?? 500)
+    await strategy.startWalk(root)
+  }
```

Comment:
Haven't read all the comment on the PR, so might have been already raised: I am not really seeing the need for this extra `newWalk()` function which just instantiates the `WalkController` and then calls `startWalk()` so I would rather stick with letting people call these consecutively themselves:

```typescript
const strategy = new WalkController()
strategy.startWalk()
```

I think I generally like this also better from a readability PoV since `newWalk()` from the naming doesn't imply yet that there is something happening (could be very well just only an instantiation function), so this can be misleading.

Reply:
I explicitly did this to ensure that our consumers will not run two walks in parallel (that is not what this class is designed for). I could rename it to `startWalk`?

Reply:
I'm assuming this is resolved now due to your new comments?

Reply:
From my side yes